cassovary-core/src/main/scala/com/twitter/cassovary/graph/labels/ArrayBasedLabel.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2015 Twitter, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+ * file except in compliance with the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package com.twitter.cassovary.graph.labels
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+import scala.reflect.runtime.universe.TypeTag
+
+trait HasMaxId {
+  val maxId: Int
+}
+
+/**
+ * A subtype of Label that uses an underlying array to map an integer id to the value
+ * associated with that id. Useful when `maxId` is not too big.
+ * @see An array based label must be defined for all ids. If desired for only some ids,
+ *      mix it in with `PartialLabel`
+ * @param name  name of the label
+ * @param maxId maximum value of the id
+ * @tparam L    type of the label value
+ */
+class ArrayBasedLabel[L : ClassTag](val name: String, val maxId: Int)(implicit val tag: TypeTag[L])
+    extends Label[Int, L] with HasMaxId {
+  private val array = new Array[L](maxId + 1)
+
+  def get(id: Int): Option[L] = Some(array(id))
```

Comment:
Actually, choosing not to specialize at all rightaway. At least in terms of storage, native arrays will be used. Will need to do performance testing before doing specialization.